[CALCITE-2067] RexLiteral cannot represent accurately floating point values, including NaN, Infinity

mihaibudiu opened the pull request:
This is another attempt to fix [CALCITE-2067]. 
PR #572 attempted to do this, but wasn't merged about 8 years ago.
This PR includes very similar changes.

The current implementation of RexLiteral uses a BigDecimal value to store DOUBLE values. This is a problem for 2 reasons:
- it loses precision, since some FP values cannot be stored precisely in decimal notation
- it cannot represent special FP values, such as -0.0, Infinity, or NaN

With this PR these issues are solved. As a side-effect, this enables the simplification engine to simplify some expressions that it couldn't before, such as expressions that produce special FP values.

Objections to the original PR which can be seen here: https://issues.apache.org/jira/browse/CALCITE-2067 include the fact that this "changes the SQL language accepted by Calcite." I want to preempt this objection: this change does not affect the parser or SqlNode, only RexNode. So it is only useful for *evaluation* and *optimization*. Moreover, if we will add support to the parser for special values as suggested in https://issues.apache.org/jira/browse/CALCITE-6058, we will certainly need a way to represent them in the RelNode IR. I claim that the changes in this PR are necessary for that goal.

The most unpleasant effect of this PR is to slightly change the semantics of some programs that used FP values before. These programs were in fact buggy, but third parties may have taken dependencies on these bugs. This kind of behavior was discovered by comparing results produced by Postgres (or Java) with Calcite on FP functions; the results were slightly off. The BigDecimal-induced rounding was the problem.

I also think that the constructor for RexLiteral that creates a literal for a DOUBLE type from a BigDecimal should be deprecated. I didn't do it in this PR.

mihaibudiu commented:
@julianhyde you had most objections to the previous PR addressing this issue. I hope that you will find my arguments adequate.

julianhyde commented on core/src/main/java/org/apache/calcite/util/Util.java:
```diff
@@ -544,6 +544,21 @@ public static void println(
+   * is in scientific notation if the value is not "special".
+   * (Special values include infinities and NaN.)
+   */
+  public static String toScientificNotation(Double d) {
```
Should this method guarantee that there is a unique string for every possible `double` value? (Except for the many representations of `NaN` etc.) And that it has an inverse method that will convert the string to the same `double` value.

I believe that we are relying on this bijection, because there is no direct way to specify a binary floating point value in SQL.

mihaibudiu replied:
Unfortunately there is no bijection. This conversion is lossy in both directions, from String to Double and from Double to String. This method is not 100% adequate for generating SQL, but it is adequate for e.g., displaying plans. 

However, the lossy conversion problem already exists in Calcite and is not specific to DOUBLE values: see https://issues.apache.org/jira/browse/CALCITE-5987. I have discussed in a message on the dev list about how this separate issue can be tackled.

mihaibudiu replied:
I realize that my previous message is wrong, indeed, there is no way to specify a binary floating point value in other SQL dialects. I don't think we can solve this problem, but we didn't introduce this problem in this PR. The problem exists already if one allows simplifications that can produce FP results.

julianhyde commented on core/src/test/java/org/apache/calcite/test/RelOptRulesTest.java:
```diff
@@ -3027,6 +3027,31 @@ private void checkPushJoinThroughUnionOnRightDoesNotMatchSemiOrAntiJoin(JoinRelT
         .check();
   }
 
+  /** Test case for <a href="https://issues.apache.org/jira/browse/CALCITE-2067">
```
Conventional format includes `[`.

julianhyde commented on core/src/main/java/org/apache/calcite/rex/RexBuilder.java:
```diff
@@ -1110,6 +1110,14 @@ public RexLiteral makeApproxLiteral(BigDecimal bd) {
     return makeApproxLiteral(bd, typeFactory.createSqlType(SqlTypeName.DOUBLE));
   }
 
+  /**
```
I think this method is too easy to invoke by accident.

mihaibudiu replied:
Yes, this should be deprecated. Can I do that?

mihaibudiu replied:
I have deleted the method

julianhyde reviewed (changes requested):
The summary "RexBuilder can't handle NaN,Infinity double constants" doesn't describe the full scope of the work. This change allows literals to have scaled binary values.

mihaibudiu commented on core/src/main/java/org/apache/calcite/rex/RexLiteral.java:
```diff
@@ -325,19 +324,18 @@ public static boolean valueMatchesType(
     case REAL:
-    case BIGINT:
-      return value instanceof BigDecimal;
+      return value instanceof BigDecimal || value instanceof Double;
```
Indeed, the decimal option should be removed, but I kept it for backward compatibility with third-party users of Calcite.

mihaibudiu replied:
I have removed the choice here too.

mihaibudiu commented:
@julianhyde I hope to have addressed your comments.
This is a breaking change in some respect; one can see how the Druid Adapter code had to be modified to accommodate this change. But I think it fixes a genuine bug. It doesn't solve the problem of FP literals completely, but it enables more programs to produce correct results than before.

rubenada commented:
bq. This is a breaking change in some respect [...]
Just a minor reminder: before the final merge, @mihaibudiu , I think it'd be necessary to include the corresponding comment about this on `history.md` in the "breaking changes" section for the next release 1.37 (so that we don't forget to mention it when preparing the next release).

mihaibudiu commented:
@rubenada I have added a commented-out breaking change for the next release. Thank you for the reminder.

mihaibudiu commented:
@julianhyde I think you have requested changes, I hope I have addressed your comments.

jduo commented on core/src/main/java/org/apache/calcite/rex/RexLiteral.java:
```diff
@@ -1060,22 +1065,46 @@ public boolean isNull() {
+        } else if (clazz == Byte.class) {
+          return clazz.cast(d.byteValue());
+        } else if (clazz == Double.class) {
+          return clazz.cast(d);
```
Isn't this just d?

mihaibudiu replied:
Yes, it is, but the return type must be T, so the cast is necessary anyway. The Java compiler isn't smart enough to understand that.

jduo commented on core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java:
```diff
@@ -388,7 +388,7 @@ private static String toSql(RelNode root, SqlDialect dialect,
     final String expected = "SELECT \"product_id\", \"shelf_width\"\n"
         + "FROM \"foodmart\".\"product\"\n"
-        + "WHERE \"net_weight\" <> 10 OR \"net_weight\" IS NULL";
+        + "WHERE \"net_weight\" <> CAST(10 AS DOUBLE) OR \"net_weight\" IS NULL";
```
Would users need to make changes to queries such as these? Is 10 no longer implicitly converted to double when comparing to double columns?

mihaibudiu replied:
no, this is just the canonical representation of the program created by Calcite which proves that the 10 is implicitly cast to a double

jduo commented on core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java:
```diff
@@ -4065,7 +4065,7 @@ private SqlDialect nonOrdinalDialect() {
         + "  from \"sales_fact_1997\") t0";
     final String expected = "SELECT SUM(CASE WHEN \"product_id\" = 0"
-        + " THEN \"net_weight\" ELSE 0 END) AS \"NET_WEIGHT\"\n"
+        + " THEN \"net_weight\" ELSE 0E0 END) AS \"NET_WEIGHT\"\n"
```
I think we'd be hard-pressed to have users use this literal FYI

mihaibudiu replied:
same as above, they don't have to.

mihaibudiu commented:
This PR is also fixing a 9-year feature request in Calcite.
From some remarks made by @julianhyde (which unfortunately I cannot find online) I deduce that he doesn't object to this change anymore.

mihaibudiu commented:
By reviewing this PR you can help close a 9-year issue for Calcite!
The changes look much bigger than they are because tests are affected, but the logic is actually very simple.

caicancai commented:
@mihaibudiu I will review this PR, before version 1.38.0, which is expected to be completed tomorrow

caicancai commented on arrow/src/main/java/org/apache/calcite/adapter/arrow/ArrowTranslator.java:
```diff
@@ -228,6 +228,8 @@ private static String getLiteralType(Object literal) {
     } else if (String.class.equals(literal.getClass())) {
       return "string";
+    } else if (literal instanceof Double) {
+      return "float";
```
why `float`?

mihaibudiu replied:
this is the same string that is used 3 lines above

caicancai commented on core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java:
```diff
@@ -1498,7 +1498,7 @@ public static SqlNode toSql(RexLiteral literal) {
       if (SqlTypeName.APPROX_TYPES.contains(typeName)) {
         return SqlLiteral.createApproxNumeric(
-            castNonNull(literal.getValueAs(BigDecimal.class)).toPlainString(), POS);
+            castNonNull(literal.getValueAs(Double.class)).toString(), POS);
```
Maybe we should add Javadoc to prevent someone from changing it back to BigDecimal later.

mihaibudiu replied:
If they change this to BigDecimal the program will crash with a type cast error.

julianhyde replied:
It might crash today, but it's possible that someone will enhance `RexLiteral.getValueAs` to accept `BigDecimal.class` as an argument. That's not an unreasonable change.

I had different concerns about this code, namely that assumes that how Java formats Double values is the same as how SQL dialects represent FLOAT literals.

I am satisfied that code is OK only because there are already quite extensive unit tests for SQL generation.

mihaibudiu replied:
BigDecimal.class should still be a legal argument for this function. 
There are a few places in the code where this happens.

This code assumes only that the createApproxNumeric will correctly handle Java string representations of double values.

caicancai commented:
Overall LGTM

julianhyde commented on core/src/test/java/org/apache/calcite/rex/RexExecutorTest.java:
```diff
@@ -389,11 +389,14 @@ interface Action {
+              rexBuilder.makeLiteral(0, integer, true));
       executor.reduce(rexBuilder, ImmutableList.of(first, second),
           reducedValues);
+      System.out.println(reducedValues);
```
Let's remove this `println`

julianhyde commented on arrow/src/test/java/org/apache/calcite/adapter/arrow/ArrowAdapterTest.java:
```diff
@@ -373,7 +373,7 @@ static void initializeArrowState(@TempDir Path sharedTempDir)
         + " where \"floatField\"=15.0";
     String plan = "PLAN=ArrowToEnumerableConverter\n"
-        + "  ArrowFilter(condition=[=(CAST($2):DOUBLE, 15.0)])\n"
+        + "  ArrowFilter(condition=[=(CAST($2):DOUBLE, 15.0E0)])\n"
```
I see a lot of `E0` added. Is it a goal of this change to make decimal literals print differently from double literals?

Maybe I missed it, but somewhere in the javadoc (e.g. for RexLiteral) should state that goal.

mihaibudiu replied:
I haven't changed how floating point values are serialized as text. The difference comes because now more literals contain a double/float.

julianhyde replied:
Makes sense.

+1 when you fix the issues in my other 2 comments

mihaibudiu commented:
I have pushed a new commit which hopefully addresses the review comments.

caicancai reviewed (approved):
LGTM

mihaibudiu commented:
This PR has been approved in JIRA, so I will merge it.
